### PR TITLE
Add a replay flag to serve files from trace directory

### PR DESCRIPTION
**Pull request by Keno**
By default rr will use the gdb `sysroot` command to load files
directly from the filesystem. However, this represents a problem
if those files get modified or deleted. This is a problem even
if the trace directory has pefectly good copies of the relevant
files (either because we managed to clone the file, or because
the trace directory got rr pack'ed). This adds a command line
flag to change that behavior and additionally adds the ability
to look through our mmap records to find a file to serve to gdb.
GDB is a bit slow at reading these files, but overall this scheme
seems to work quite well.

### CODE REVIEW

**Comment by rocallahan** [comment]
The boost/std-fs dependency worries me quite a bit. How much harder would it be to use regular syscalls?

**Comment by rocallahan** [comment]
Keep in mind that people are still building rr on ludicrously old systems.

**Comment by Keno** [comment]
It's just a bunch of annoying string wrangling that I didn't want to get into if there was a perfectly good library, but should be doable - I'll look into it.

**Comment by Keno** [comment]
Fixed with just doing the minimal path manipulation and not doing all the corner cases the standard mandates. I think that's likely to be fine for our use case.

File: src/GdbServer.cc [hunk 1]
```diff
@@ -1741,7 +1794,49 @@ static ScopedFd generate_fake_proc_maps(Task* t) {
...
+  string fname = map_name.substr(matchpos == string::npos ? 0 : matchpos + 1);
+  return memcmp(fname.c_str(), ld_start,
+                sizeof(ld_start)-1) == 0;
+}
+
+static bool is_likely_interp(std::string fsname) {
+  return fsname == "/lib64/ld-linux-x86-64.so.2" || fsname == "/lib/ld-linux.so.2";
+}
+
+std::string normalize_path(std::string path)
```
Comment (rocallahan) [inline src/GdbServer.cc:1809]: Please document what normalization is done.

File: src/GdbServer.cc [hunk 2]
```diff
@@ -1423,7 +1473,7 @@ struct DebuggerParams {
   short port;
 };
 
-static void push_default_gdb_options(vector<string>& vec) {
+static void push_default_gdb_options(vector<string>& vec, bool serve_files=false) {
```
Comment (rocallahan) [inline src/GdbServer.cc:1476]: spaces around =

File: src/GdbServer.cc [hunk 3]
```diff
@@ -1741,7 +1794,49 @@ static ScopedFd generate_fake_proc_maps(Task* t) {
   return move(file.fd);
 }
 
-int GdbServer::open_file(Session& session, const std::string& file_name) {
+static bool is_ld_mapping(std::string map_name) {
```
Comment (rocallahan) [inline src/GdbServer.cc:1797]: I think we're `using namespace std` here so we can avoid the `std::`

File: src/GdbServer.cc (hunk 1 above)
Comment (rocallahan) [reply src/GdbServer.cc:1809]: Can we merge this with `normalize_file_name` in `SourcesCommand.cc`? I apologise for not documenting that properly!

File: src/GdbServer.h [hunk 4]
```diff
@@ -207,7 +208,7 @@ class GdbServer {
    * an entry to `files` with the file contents and return our internal
    * file descriptor.
    */
-  int open_file(Session& session, const std::string& file_name);
+  int open_file(Session& session, Task *continue_task, const std::string& file_name);
```
Comment (rocallahan) [inline src/GdbServer.h:211]: Space after * not before

**Pull request merged**